dom/src/DOMSourceFactory.ts

```diff
@@ -0,0 +1,24 @@
+import {DOMSource} from './DOMSource';
+import {SourceOptions} from './SourceOptions';
+import {MainDOMSource, MainDOMSourceOptions} from './MainDOMSource';
+import {BodyDOMSource} from './BodyDOMSource';
+import {DocumentDOMSource} from './DocumentDOMSource';
+
+export class DOMSourceFactory {
+  static create(type: string, options: DOMSourceOptions): DOMSource {
+    // Ordered most common on top for optimisation.
+    // We use if-statements instead of switch, because few conditionals
+    // optimise better with if-statements.
+    if (type !== 'document' && type !== 'body') {
+      return new MainDOMSource(<MainDOMSourceOptions>options);
+    }
+    if (type === 'body') {
+      return new BodyDOMSource(options);
+    }
+    if (type === `document`) {
+      return new DocumentDOMSource(options);
+    }
+  }
+}
+
+export type DOMSourceOptions = SourceOptions | MainDOMSourceOptions;
```

Comment:
Ok, there is some weirdness here. 
You don't need `DOMSourceOptions = SourceOptions | MainDOMSourceOptions`, you could just use `SourceOptions` because `MainDOMSourceOptions extends SourceOptions`, so "SourceOptions or MainDOMSourceOptions" is always covered by just "SourceOptions".
Then, I don't like the name "SourceOptions", I think it should be "DOMSourceOptions". "Source" is too generic, might sound like it's coming from `@cycle/base`.